### PR TITLE
Minor tweak to GitHub Actions

### PR DESCRIPTION
**Pull request by dra27**
This PR overhauls our CI system to migrate the Windows testing from AppVeyor to GitHub Actions. With the caches primed, the macOS test job appears to be the slow step with the entire run taking 40-45 minutes.

The change-set is very big! Although GitHub has increased the cache limits since I originally started prototyping this last summer, it's still the case that adding the Windows compilers and Cygwin puts tremendous pressure on our cache quotas. This PR necessarily includes an overhaul of non-Windows parts as a result. I've also included - for better or worse (probably the latter) - a rudimentary generator for the YAML workflows. In its defence, it found some bugs. The original version of this PR had macOS, Windows and Linux as separate workflow files, making the case for a generator blindingly obvious. The main thing it's trying to do is ensure that the caches are used consistently (keys, paths, etc.) and also that the jobs which get fanned out over multiple platforms are identical. It does also allow platforms and versions to be added relatively quickly if they're needed for some reason. Critically, it's entirely possible - and, indeed, how I did it - to hack `main.yml`, commit those changes and only _then_ update `ci.ml` to ensure the changes made have been done consistently.

I haven't turned off AppVeyor in this PR yet, as strictly speaking we're not yet doing _exactly_ the same test matrix, but that can follow very soon after.

Headline changes:
- actions/cache contains some critical flaws (all of which are known to GitHub, most of which have been contributed as fixes and all of which have not been merged!). In particular, reliably sharing caches between Unix and Windows requires forcing those caches to be saved using gzip, not zstd. Especially on Windows, but noticeably on Unix as well, it's desirable in the early stages to be able to test for the presence of a cache, but not download it - i.e. where you just want to know if the stage is required. As a result, I've forked actions/cache and added two trivial parameters to it `check-only` and `force-gzip`. It's a couple of commits at https://github.com/dra27/cache/commits/opam and is trivial to rebase when actions/cache is updated (it seems to be updated quite infrequently). I suggest that we move the commits to a fork in ocaml-opam. Again, it _significantly_ reduces run times when the caches are actually primed. Looking at the (cached) run of this PR in [GitHub Actions](https://github.com/dra27/opam/actions/runs/1822506039), you can see the Analyse and Cygwin job taking < 10s where they can take over a minute if they actually extract the files.
- macOS, Windows and Linux are tested in their own matrices. The idea here is that Linux tends to complete quickly and you actually get feedback on the success of your PR within roughly 10 minutes (including tests). macOS and Windows then catch up and confirm that you haven't broken them!
- The main compiler caches now also cache dune, which means we don't test dune every time we test opam! It's a noticeable time saving, even on Linux (look at the time for Build step in the Build-Linux matrix)
- A little dance is done to cache the `_build` directory for the src_exts as well. Again, there's quite a build time benefit to this, especially on Windows. The trickery here is that we erase all the rest of opam from `_build` and dune then figures out what to do on the next invocation for each PR. This could be extended further to cache `_build` for actual PRs, but that's fiddly (and likely to risk blowing cache quotas)

This is best reviewed commit-by-commit and certain with a diff'ing tool which can ignore whitespace changes as much as possible. The more detailed changes are:
- Propagate the use of `src_ext/archives` cache further around the workflow (pre-existing issue). Speeds up src_ext compilation and improves consistency as we're less exposed to network access.
- The opam-repository clone is included in the commit, with some attempt at ensuring that if a different fork is temporarily used then the caches are ignored. Again limits network access and chance for failure.
- `ocaml-secondary-compiler` is cached for legacy jobs, greatly speeding them up. It's alas a bit too awkward to cache `dune` as well, but it doesn't take as long.
- The test stages were using a separate compiler which had `ocamldoc`. This was for Zarith which erroneously requires it. I've submitted a PR for that upstream but worked around it here by installing a script for `ocamldoc` which returns an error. This allows all the builds on Linux to use the _same_ compiler more than halving the cache usage for those ocaml versions.
- Cygwin includes `curl` by default, not `wget`, so I've switched CI to prefer `curl` to `wget`.
- There's a minor fix - not used in CI - but which allows the legacy build matrix to be run on macOS if needed.
- Passing things around via the environment is both a bit confusing given the number of scripts we have but also quite tricky on Windows. I've eliminated a few of them in preference for explicit command line arguments - in particular, `OCAML_VERSION` is gone.
- There's a tiny bug in `make -C src_ext cache-archives` which is fixed... technically it wasn't downloading lib-pkg-only archives before.
- The test stages are split up to make them a little more obvious in the logs but also to be able to skip them on Windows more easily, etc.

### CODE REVIEW

File: .github/scripts/main/ocaml-cache.sh [hunk 1]
```diff
@@ -2,43 +2,152 @@
...
 
 cd ..
 rm -rf "ocaml-$OCAML_VERSION"
+
+if [[ $PLATFORM != 'Windows' ]]; then
+  echo > "$OCAML_LOCAL/bin/ocamldoc" <<"EOF"
+#!/bin/sh
+
+echo 'ocamldoc is not supposed to be called'>&2
+exit 1
```
Comment (rjbou) [inline .github/scripts/main/ocaml-cache.sh:128]: *joli!*

**Comment by kit-ty-kate** [comment]
should we maybe remove the x86_32? They seem a bit redundant with the windows x86_64

**Comment by rjbou** [comment]
Some test on the PR (i'll update this comment at each new test):

| test | status | time | note |
| -- | -- | -- | -- |
| [all green with no cache](https://github.com/rjbou/opam/actions/runs/1846688237) | :heavy_check_mark:  | 1h12 | |

## Compilation

| test | status | time | note |
| -- | -- | -- | -- |
| [failing for all](https://github.com/rjbou/opam/actions/runs/1847587460)     | :hourglass_flowing_sand:   | 
| [failing for old compiler]() | :hourglass_flowing_sand: | 

## Reftests

| test | status | time | note |
| -- | -- | -- | -- |
| [failing for all](https://github.com/rjbou/opam/actions/runs/1847071261)        | :heavy_check_mark:  | 49m
| [failing on linux]() | :hourglass_flowing_sand:  |
| [failing on windows]() | :hourglass_flowing_sand: |
| [failing on macos]() | :hourglass_flowing_sand: |

## Errors
* [ ] Hygiene: The head commit for this push event is not ahead of the base commit. Please submit an issue on this action's GitHub repo.

**Review by kit-ty-kate — APPROVED** [review]

**Comment by rjbou** [comment]
finished testng it!

**Comment by rjbou** [comment]
Add patches & cli tests todo

**Comment by dra27** [comment]
I think the `configure` check is fixed by https://github.com/ocaml/opam/pull/5055/commits/baecf33147f77df80b3eb21d845316c89aaf1c83?

I just pushed an extra commit which tweaks the way the `src_ext` builds are cached with the compiler. You can see from [this run](https://github.com/dra27/opam/runs/5499024487?check_suite_focus=true) that a broken build fails in the `Build` step, even if the compiler had to be built.

**Comment by kit-ty-kate** [comment]
The `*-pc-cygwin` jobs fail with this now, for some reason:
```
    ocamlopt src/client/opamMain.exe (exit 2)
(cd _build/default && /cygdrive/d/Cache/ocaml-local/bin/ocamlopt.opt -w -40 -w +a-4-40-42-44-48 -safe-string -g -o src/client/opamMain.exe src_ext/seq/src/seq.cmxa src_ext/re/lib/re.cmxa src_ext/stdlib-shims/src/stdlib_shims.cmxa src_ext/ocamlgraph/src/graph.cmxa /cygdrive/d/Cache/ocaml-local/lib/ocaml/unix.cmxa -I /cygdrive/d/Cache/ocaml-local/lib/ocaml /cygdrive/d/Cache/ocaml-local/lib/ocaml/bigarray.cmxa -I /cygdrive/d/Cache/ocaml-local/lib/ocaml src/stubs/libacl/opam_stubs_libacl.cmxa -I src/stubs/libacl src/core/opam_core.cmxa src_ext/opam-file-format/src/opam_file_format.cmxa src/format/opam_format.cmxa src/repository/opam_repository.cmxa src_ext/spdx_licenses/src/spdx_licenses.cmxa src/state/opam_state.cmxa src_ext/extlib/src/extlib.cmxa src_ext/cudf/cudf.cmxa src_ext/re/deprecated/re_pcre.cmxa src_ext/dose3/src/common/dose_common.cmxa src_ext/dose3/src/algo/dose_algo.cmxa src_ext/0install-solver/src/solver/zeroinstall_solver.cmxa src_ext/opam-0install-cudf/lib-cudf/opam_0install_cudf.cmxa src_ext/mccs/src/glpk/mccs_glpk.cmxa -I src_ext/mccs/src/glpk src_ext/mccs/src/mccs.cmxa -I src_ext/mccs/src src/solver/opam_solver.cmxa -I /cygdrive/d/Cache/ocaml-local/lib/ocaml src_ext/base64/src/base64.cmxa src_ext/cmdliner/src/cmdliner.cmxa src/client/opam_client.cmxa src/client/.opamMain.eobjs/native/dune__exe__OpamMain.cmx)
** Cannot resolve symbols for src_ext/mccs/src/libmccs_stubs.a(constraint_generation.o/
count_criteria.o/
cudf_reductions.o/
lexagregate_combiner.o/
notuptodate_criteria.o/
removed_criteria.o/
):
 __Unwind_Resume
** Cannot resolve symbols for src_ext/mccs/src/libmccs_stubs.a(glpk_solver.o):
 __Unwind_Resume
** Cannot resolve symbols for src_ext/mccs/src/libmccs_stubs.a(lp_solver.o):
 __Unwind_Resume
** Cannot resolve symbols for src_ext/mccs/src/libmccs_stubs.a(mccs_stubs.o):
 __Unwind_Resume
** Cannot resolve symbols for src_ext/mccs/src/libmccs_stubs.a(cudf_tools.o):
 __Unwind_Resume
** Cannot resolve symbols for src_ext/mccs/src/libmccs_stubs.a(mccscudf.o):
 __Unwind_Resume
** Cannot resolve symbols for src_ext/mccs/src/libmccs_stubs.a(cudf_reductions.o/
lexagregate_combiner.o/
notuptodate_criteria.o/
removed_criteria.o/
):
 __Unwind_Resume
File "caml_startup", line 1:
Error: Error during linking (exit code 2)
make: *** [Makefile:131: build-opam] Error 1
Error: Process completed with exit code 2.
```

**Comment by dra27** [comment]
Ah, I didn't predict that, but that's strangely correct - it's revealing the fact that the master branch is presently broken for cygwin32! I'll split https://github.com/ocaml/opam/pull/5055/commits/8d45c76071cfae1cb4652a7a4df874a43c42ad3a off to a separate PR - I don't think it's worth making the cache logic harder, because once this is merged that should be a more obscure corner case: this case happens when the src_ext build on the branch you are merging into is broken **and** the cache of the compiler is missing.

**Comment by rjbou** [comment]
> I think the configure check is fixed by https://github.com/ocaml/opam/commit/baecf33147f77df80b3eb21d845316c89aaf1c83?

no, it just fixes the GH yaml file generation check, not the configure issue. Even for ok configure changes, the configure check is failing. But I can't say if it's job issue or autoconf issue

**Pull request merged**